Nested queries revamp base

sbelak opened the pull request:
This lays the groundwork for nested queries (and sanboxing) fixes. Fundamentally it does 4 (related) things:
1) allows `field-id` to be used for fields from a source query
2) improves how we flow metadata for fields
3) cleans up how we generate SQL for queries with expressions 
4) changes our SQL generation to decorate fields with table name prefix 

## Limitations

* We still don't handle well the case where source query has two joins to the same table that both pull in the same fields. This is not so much a limitation of the current approach as it is of MBQL semantics. The problem is we don't have a good way to reference such fields. This is one area where the existing approach with `field-literals` was superior. It however relied on an implementation detail of how we disambiguate duplicate names (by appending __N) which in itself is quite brittle as it's position-dependent. This means that small changes to the source query could cause the order of how we process and name fields change, and break things (or alternatively it places another invariant on the query processing pipeline that ordering has to be stable). In addition it also means more work for the frontend as dropping columns can mean that `field-literal` references become invalid. I think that the best long-term solution is to introduce an options argument to `field-id` which could be used to disambiguate via something like `:source_alias` clause.  

* The metadata flowing is currently aggressive only for fields because we can be aggressive about matching via ID and it's the most common case of clashes. The approach taken should work just as well for things referred to by `field-id`, but is something I would attempt only after (& if) we implement `:source_alias` clause from above. Issues here can only arise if you do something silly like using the same expression name for multiple different expressions; or shadowing field names with expression names or aggregation names; etc. 

* Table name prefix could be insufficient if you do weird stuff and join from tables with the same name but different schemas. I was unsure whether to make the names even uglier to work around that. We could easily modify `unambiguous-field-alias` to either include the schema, or change it to a hash of table name + table schema, or similar.

- Fixes: 
    - #12649, ✅
    - #12762, ✅
    - metabase/metabase-enterprise#520,
    - metabase/metabase-enterprise#154,
    - metabase/metabase-enterprise#524,
    ... [more to be added as I put back the tests]

sbelak committed:
Fix SQL for custom columns referring to aggregations. Fixes #12762

sbelak committed:
Add test & enable cypress

sbelak committed:
Disambiguate fields with same name in source query

sbelak committed:
Use field-id where possible

sbelak committed:
More untaglment

sbelak committed:
move defs around

sbelak committed:
lift joined fields when lifting expressions

sbelak committed:
cleanup

sbelak committed:
More cleanup

sbelak committed:
More cleanup

sbelak committed:
Merge branch 'master' into fix-nested-questions-with-joins-with-name-clashes-12649

sbelak committed:
Reuse alias in unique name

sbelak committed:
Add `source_alias` to metadata map

sbelak committed:
Remove cruft

sbelak committed:
Simplify expression generation

sbelak committed:
Don't try to infer col types of native queries

sbelak committed:
Fix metadata flowing for field ids

sbelak committed:
Remove unneeded require

sbelak committed:
Fix tests

sbelak commented on src/metabase/query_processor/middleware/annotate.clj:
```diff
@@ -58,10 +58,9 @@
   "Double-check that the *driver* returned the correct number of `columns` for native query results."
-  [cols rows]
-  {:pre [(sequential? cols) (every? map? cols)]}
+  [cols :- [{s/Any s/Any}], rows]
```
This felt cleaner as we want to also support the version where `cols` = nil

sbelak committed:
Banish mbql->honeysql

sbelak committed:
Correctly infer base type for expressions field-literal

sbelak committed:
Fix tests

sbelak committed:
Merge branch 'master' into nested-queries-revamp-base

sbelak committed:
More test fixes

rlotun commented:
just a reminder that there is a reproduction for bug this issue closes - see info at https://github.com/metabase/metabase/pull/13557

sbelak committed:
Fix BQ driver

sbelak committed:
Handle joined fields bette

sbelak commented on test/metabase/query_processor_test/explicit_joins_test.clj:
```diff
@@ -387,22 +380,24 @@
                            {source-query :query} :dataset_query
                            source-metadata       :result_metadata} (qp.test-util/card-with-source-metadata-for-query
-                                                                    (mt/mbql-query checkins
+                                                                    (mt/mbql-query venues
```
I've changed this query because the self join was confusing and was detracting from what's really the point.

sbelak commented on src/metabase/driver/sql/query_processor.clj:
```diff
@@ -301,7 +312,8 @@
 (defmethod ->honeysql [:sql :joined-field]
   [driver [_ alias field]]
-  (binding [*table-alias* alias]
+  (binding [*table-alias*   alias
```
This is ugly and it's making me sad. I'm torn between using a flag and introducing `*join-alias*`. In the latter case `->honeysql` for `:joined-field` would only set `*join-alias*`, and we'd use `(or *joined-alias* *source-alias)` where relevant.

Adjacent to the issues around `:source_alias` I mention above.

sbelak committed:
Merge branch 'master' into nested-queries-revamp-base

sbelak committed:
Remove unneeded requires

sbelak commented:
The EE test that's failing requires #13715 to be done

camsaul reviewed (approved):
This works for me

rlotun commented:
can we publish this PR? it looks like we have an approved draft PR, which seems a bit strange

sbelak committed:
Merge branch 'master' into nested-queries-revamp-base

sbelak committed:
Update test to mt

sbelak committed:
simplify test

sbelak committed:
Merge branch 'master' into nested-queries-revamp-base

sbelak committed:
Unmangle merge

sbelak committed:
Add cypress tests

sbelak committed:
Merge branch 'master' into nested-queries-revamp-base

sbelak committed:
Remove unneeded requires

sbelak committed:
Merge branch 'master' into nested-queries-revamp-base

sbelak committed:
Get metadata for all fields used not just result cols

sbelak committed:
typo

sbelak committed:
use mt/format-name for `source_alias` in tests

sbelak commented:
Oracle before 12.2 apparently doesn't like identifier names > 30. Do we sacrifice some readability of the generated SQL and just use a hash prefix, or do we do some driver-specific truncation?

sbelak committed:
Explicitly use uppercasing when normalizing `:source_alias`

sbelak commented:
Probably makes more sense to push the name limit to the drivers so we also fix #9204. With Oracle the issue is somewhat more pernicious though as it comes up during our regular tests and if we mangle names differently for different dbs a lot of tests will break.

sbelak committed:
Oracle: truncate identifiers if needed

sbelak committed:
Correctly use Identifier

sbelak committed:
Operate only on the last part of the identifier

sbelak committed:
add missing arg

sbelak committed:
Oracle: Use only letters in identifier names

sbelak committed:
Normalize source_alias in tests

nemanjaglumac commented:
[metabase/metabase-enterprise#520](https://github.com/metabase/metabase-enterprise/issues/520) is still failing:

![image](https://user-images.githubusercontent.com/31325167/105099530-1ea5f700-5aac-11eb-9ea5-58d76eba8142.png)

nemanjaglumac commented:
[metabase/metabase-enterprise#154](https://github.com/metabase/metabase-enterprise/issues/154) is still failing:

![image](https://user-images.githubusercontent.com/31325167/105099809-8c522300-5aac-11eb-954c-6b2ee3ad3789.png)

sbelak committed:
Bigquery: update hardcoded query

nemanjaglumac committed:
Merge branch 'master' into nested-queries-revamp-base

nemanjaglumac committed:
Delete duplicate test

Note: this is a left-over from before. I missed it somehow. Removing it now to avoid further confusion.

nemanjaglumac committed:
Update repro for #12928 [ci skip] (#14468)

Expose error in Cypress runner rather than relying on UI.

rlotun commented on modules/drivers/oracle/src/metabase/driver/oracle.clj:
```diff
@@ -127,6 +129,30 @@
+       (map (fn [digit]
+              (-> digit
+                  int
+                  (+ 65)
```
where does the 65 come from?

camsaul replied:
65 is ASCII `A` so I'm guess Simon is converting the digits in the truncated identifier here to letters

rlotun replied:
yeah, realized after ;-)

rlotun reviewed (changes requested):
1. Seems like we need to unskip the test for EE#520: https://github.com/metabase/metabase/commit/e1e4278a4ee3b1487b84a32f1197f87865b9e6d0#diff-0b620ecf37aabe49d79827f61977722016f157674b390a23f28e1ac1096b8a79R779-R795

2. We should also unskip the tests for EE#154: https://github.com/metabase/metabase/commit/394dfa83f37628118a09f286e74eb3ff037ec73d#diff-0b620ecf37aabe49d79827f61977722016f157674b390a23f28e1ac1096b8a79R278-R284

~~3. We need to unskip tests for EE#524: https://github.com/metabase/metabase/pull/13902/files#diff-0b620ecf37aabe49d79827f61977722016f157674b390a23f28e1ac1096b8a79R577-R584~~

sbelak commented:
FK remappings in nested queries/sandboxes now work again (part of ee/520 regression).

sbelak committed:
Move sandobxing middleware

nemanjaglumac commented:
> FK remappings in nested queries/sandboxes now work again (part of ee/520 regression).

Can you please unskip EE_520 to make sure this indeed works. Also, if there are any other related issues that may be fixed by this, they should be unskipped as well.

sbelak commented:
It fixes one half or the issues in 520? Should I split the cypress test in two so each can be done individually or just wait until I have the full fix

nemanjaglumac commented:
> It fixes one half or the issues in 520? Should I split the cypress test in two so each can be done individually or just wait until I have the full fix

No. Please don't. Let's wait for the full fix in that case 💪

sbelak committed:
Make field read permissions work with sandboxing

sbelak committed:
Sandboxing: use table metadata for native source cards

sbelak committed:
Re-enable cypress tests for ee/520 & ee/154

sbelak committed:
Add missing arg

sbelak commented:
Fixed the regressions and re-enabled cypress tests for ee/520 & ee/154. Everything green
